mac/Focalboard/ViewController.swift

```diff
@@ -64,20 +71,55 @@ class ViewController:
 	@objc func onServerStarted() {
 		NSLog("onServerStarted")
 		DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
-			self.updateSessionToken()
+			self.updateSessionTokenAndUserSettings()
 			self.loadHomepage()
 		}
 	}
 
-	private func updateSessionToken() {
+	private func persistUserSettings() {
+		let semaphore = DispatchSemaphore(value: 0)
+
+		webView.evaluateJavaScript("Focalboard.exportUserSettingsBlob();") { result, error in
+			defer { semaphore.signal() }
+			guard let blob = result as? String else {
+				NSLog("Failed to export user settings: \(error?.localizedDescription ?? "?")")
+				return
+			}
+			UserDefaults.standard.set(blob, forKey: "localStorage")
+			NSLog("Persisted user settings: \(Data(base64Encoded: blob).flatMap { String(data: $0, encoding: .utf8) } ?? blob)")
+		}
+
+		// During shutdown the system grants us about 5 seconds to clean up and store user data
+		let timeout = DispatchTime.now() + .seconds(3)
+		var result: DispatchTimeoutResult?
+
+		// Busy wait because evaluateJavaScript can only be called from *and* signals on the main thread
+		while (result != .success && .now() < timeout) {
+			result = semaphore.wait(timeout: .now())
+			RunLoop.current.run(mode: .default, before: Date())
+		}
+
+		if result == .timedOut {
+			NSLog("Timed out trying to persist user settings")
+		}
+	}
+
+	private func updateSessionTokenAndUserSettings() {
 		let appDelegate = NSApplication.shared.delegate as! AppDelegate
-		let script = WKUserScript(
+		let sessionTokenScript = WKUserScript(
```

Comment:
if we are populating localstorage from the user preferences, should we first clear it so there are no old elements getting in the way (like an upgrade/downgrade?)

Reply:
Hm, good question. I wonder if that's not a general problem that the web app needs to handle? When you use Focalboard in the browser and switch to a new version, it could happen as well that `localStorage` contains data from previous versions. If the semantics of the user settings change between versions, some migration logic will be needed but I'm not sure if we need to do anything special for this from the native app side. If we insert the last known settings into `localStorage` on launch, it's essentially the same problem as on the web version from that point on.